### PR TITLE
Feature/86 signin form validation

### PR DESCRIPTION
**Pull request by SiboYang**
# Summary

- Use RegEx to test for email input, if not valid, a error message will be displayed, no error message when the no email is entered.
- checked the length of password, error message will be displayed if password is shorter than 8 (github password is at least 8 charaters long, no sure if we will use the same standard).
- Sign in button is only available when email and password are both valid.
- Error message (if any) will be displayed after clicked sign in. ("Network error", "Incorrect password", "No user record corresponding to this email")

## Test Plan

Manually

## Related Issues

#86

### CODE REVIEW

File: frontend/src/App/App.tsx [hunk 1]
```diff
@@ -7,37 +7,52 @@ import PlusButton from "../plusButton/PlusButton";
...
+        setSignInError('')
+      })
+      .catch((error: any) => {
+        console.log(error.message);
+        if (error.code === "auth/network-request-failed") {
+          setSignInError("Network error.")
+        } else if (error.code === 'auth/wrong-password') {
+          setSignInError("Incorrect password.")
+        } else if (error.code === "auth/user-not-found") {
+          setSignInError("No user record corresponding to this email.")
```
Comment (tedspare) [inline frontend/src/App/App.tsx:36]: nice error messages!

these could be combined to a switch statement but it's fairly readable now

File: frontend/src/App/App.tsx [hunk 2]
```diff
@@ -7,37 +7,52 @@ import PlusButton from "../plusButton/PlusButton";
...
     FirebaseAuth.onAuthStateChanged((user: any) => {
-      setUser(user)
-      setIsSignedIn(!!user)
-      setSigningIn(false)
+      setUser(user);
+      setIsSignedIn(!!user);
+      setSigningIn(false);
     });
-  }, [])
+  }, []);
```
Comment (tedspare) [inline frontend/src/App/App.tsx:51]: I'll add semicolons in the future to avoid this lol;

File: frontend/src/SignInForm/SigninForm.tsx [hunk 3]
```diff
@@ -39,12 +41,22 @@ const useStyles = makeStyles((theme: Theme) =>
...
-  const [email, setEmail] = useState<string>("")
-  const [password, setPassword] = useState<string>("")
+  const [email, setEmail] = useState<string>("");
+  const [password, setPassword] = useState<string>("");
+  const [emailIsValid, setEmailIsValid] = useState<boolean>(false);
+  const [passwordIsValid, setPasswordIsValid] = useState<boolean>(false);
 
   const classes = useStyles();
+  const emailPattern = new RegExp(
+    /^(("[\w-\s]+")|([\w-]+(?:\.[\w-]+)*)|("[\w-\s]+")([\w-]+(?:\.[\w-]+)*))(@((?:[\w-]+\.)*\w[\w-]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$)|(@\[?((25[0-5]\.|2[0-4][0-9]\.|1[0-9]{2}\.|[0-9]{1,2}\.))((25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})\.){2}(25[0-5]|2[0-4][0-9]|1[0-9]{2}|[0-9]{1,2})\]?$)/i
```
Comment (tedspare) [inline frontend/src/SignInForm/SigninForm.tsx:58]: I can't read regex but do two-letter TLDs like "sibo@angular.io" work?

do dots and dashes like "sibo.yang-backup@email.com" work?

**Review by tedspare — APPROVED** [review]
LGTM

File: frontend/src/App/App.tsx (hunk 1 above)
Comment (SiboYang) [reply frontend/src/App/App.tsx:36]: They both work
Screen Shot 2021-01-27 at 7 30 04 PM
Screen Shot 2021-01-27 at 7 30 19 PM

File: frontend/src/App/App.tsx (hunk 2 above)
Comment (SiboYang) [reply frontend/src/App/App.tsx:51]: lol

File: frontend/src/SignInForm/SigninForm.tsx (hunk 3 above)
Comment (SiboYang) [reply frontend/src/SignInForm/SigninForm.tsx:58]: They both work
Screen Shot 2021-01-27 at 7 30 04 PM
Screen Shot 2021-01-27 at 7 30 19 PM

**Pull request merged**